reasoner/resolution/resolver/BoundConclusionResolver.java

```diff
@@ -0,0 +1,377 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.reasoner.resolution.resolver;
+
+import com.vaticle.typedb.common.collection.Pair;
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+import com.vaticle.typedb.core.common.iterator.FunctionalIterator;
+import com.vaticle.typedb.core.common.iterator.Iterators;
+import com.vaticle.typedb.core.concept.Concept;
+import com.vaticle.typedb.core.concept.ConceptManager;
+import com.vaticle.typedb.core.concept.answer.ConceptMap;
+import com.vaticle.typedb.core.logic.Rule;
+import com.vaticle.typedb.core.reasoner.resolution.ResolverRegistry;
+import com.vaticle.typedb.core.reasoner.resolution.answer.AnswerState;
+import com.vaticle.typedb.core.reasoner.resolution.answer.AnswerState.Partial.Concludable;
+import com.vaticle.typedb.core.reasoner.resolution.framework.Request;
+import com.vaticle.typedb.core.reasoner.resolution.framework.RequestState;
+import com.vaticle.typedb.core.reasoner.resolution.framework.ResolutionTracer;
+import com.vaticle.typedb.core.reasoner.resolution.framework.Resolver;
+import com.vaticle.typedb.core.reasoner.resolution.framework.Response;
+import com.vaticle.typedb.core.traversal.GraphTraversal;
+import com.vaticle.typedb.core.traversal.TraversalEngine;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class BoundConclusionResolver extends Resolver<BoundConclusionResolver> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BoundConclusionResolver.class);
+    private final ConceptMap bounds;
+    private final Driver<ConditionResolver> conditionResolver;
+    private final Rule.Conclusion conclusion;
+    private final Driver<Materialiser> materialiser;
+    private final Map<Request, ConclusionRequestState<? extends Concludable<?>>> requestStates;
+    private final Map<Materialiser.Request, Pair<Request, Integer>> materialiserRequestRouter;
+
+    public BoundConclusionResolver(Driver<BoundConclusionResolver> driver, Rule.Conclusion conclusion, ConceptMap bounds,
+                                   Driver<ConditionResolver> conditionResolver,
+                                   Driver<Materialiser> materialiser, ResolverRegistry registry, TraversalEngine traversalEngine,
+                                   ConceptManager conceptMgr, boolean resolutionTracing) {
```

Comment:
Ok true that approach makes sense. I have been cautious because I don't want concurrent modification of the registry, but as long as that's avoided we're ok.

Reply:
In fact I now wonder this: if the Bound resolvers are to be orchestrated by "coordinators", then effectively those coordinators are registries in the same way as the `ResolverRegistry`. In which case they should all become the same class of thing, and the `ResolverRegistry` should be an actor too.